plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaHttpClient.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.openpolicyagent;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.FluentFuture;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.Inject;
+import io.airlift.http.client.FullJsonResponseHandler;
+import io.airlift.http.client.HttpClient;
+import io.airlift.http.client.HttpStatus;
+import io.airlift.http.client.Request;
+import io.airlift.json.JsonCodec;
+import io.trino.plugin.openpolicyagent.schema.OpaBatchQueryResult;
+import io.trino.plugin.openpolicyagent.schema.OpaQuery;
+import io.trino.plugin.openpolicyagent.schema.OpaQueryInput;
+import io.trino.plugin.openpolicyagent.schema.OpaQueryResult;
+
+import java.net.URI;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static com.google.common.net.MediaType.JSON_UTF_8;
+import static io.airlift.http.client.FullJsonResponseHandler.createFullJsonResponseHandler;
+import static io.airlift.http.client.JsonBodyGenerator.jsonBodyGenerator;
+import static io.airlift.http.client.Request.Builder.preparePost;
+import static java.util.stream.Collectors.toMap;
+
+public class OpaHttpClient
+{
+    private final HttpClient httpClient;
+    private final JsonCodec<OpaQuery> serializer;
+    private final Executor executor;
+
+    @Inject
+    public OpaHttpClient(
+            @ForOpa HttpClient httpClient,
+            JsonCodec<OpaQuery> serializer,
+            @ForOpa Executor executor)
+    {
+        this.httpClient = httpClient;
+        this.serializer = serializer;
+        this.executor = executor;
+    }
+
+    public <T> FluentFuture<T> submitOpaRequest(OpaQueryInput input, URI uri, JsonCodec<T> deserializer)
```

Comment:
If requests fail to send, we'll throw an exception. Any non-200 response from OPA is dealt with in `parseOpaResponse`.

Failed requests are not retried, they are treated in the same way as a deny

Reply:
In my experience HTTP request fail all the time for various, spurious, reasons.  I suggest adding retries in the future for anything that isn't the server actively saying "I understand what you are asking and access is denied".  This will work around problems like online upgrades where the service temporarily gets redirected to a server that returns 404 for example.  This doesn't need to be done in this commit, but I'd put it on the todo list.